### PR TITLE
feat(compiler): allow to create ChangeDetectors from parsed templates

### PR DESCRIPTION
**Pull request by tbosch**
Part of #3605

### CODE REVIEW

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
+import {Injectable} from 'angular2/di';
+
+import {List, ListWrapper, MapWrapper} from 'angular2/src/core/facade/collection';
```
Comment (gkalpak) [inline modules/angular2/src/compiler/change_definition_factory.ts]: Is `List` needed ?

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    var directiveIndex;
+    if (isBlank(prevDirRecord)) {
+      directiveIndex = new DirectiveIndex(this.currentProtoView.boundElementCount - 1, 0);
+    } else {
+      directiveIndex = new DirectiveIndex(prevDirRecord.directiveIndex.elementIndex,
+                                          prevDirRecord.directiveIndex.directiveIndex + 1);
+    }
+    var directiveMetadata = ast.directive;
+    var changeDetectionMeta = directiveMetadata.changeDetection;
+    this.directiveRecord = new DirectiveRecord({
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: callbacks are missing:
```
            callAfterContentInit: directiveMetadata.callAfterContentInit,
            callAfterContentChecked: directiveMetadata.callAfterContentChecked,
            callAfterViewInit: directiveMetadata.callAfterViewInit,
            callAfterViewChecked: directiveMetadata.callAfterViewChecked,
            callOnChanges: directiveMetadata.callOnChanges,
            callDoCheck: directiveMetadata.callDoCheck,
            callOnInit: directiveMetadata.callOnInit,
```

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  var visitor = new ProtoViewVisitor(componentStrategy);
+  templateVisitAll(visitor, parsedTemplate);
+  return createChangeDefinitions(visitor.protoViewStack, componentType, genConfig);
+}
+
+class ProtoViewVisitor implements TemplateAstVisitor {
+  viewCount: number = 0;
+  currentProtoView: ProtoViewVisitorData;
+  protoViewStack: ProtoViewVisitorData[] = [];
+  result: ChangeDetectorDefinition[] = [];
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: not needed, remove

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    this.viewCount++;
+    var prevStack = this.currentProtoView;
+    this.currentProtoView =
+        new ProtoViewVisitorData(prevStack, ChangeDetectionStrategy.Default, this.viewCount - 1);
+    this.protoViewStack.push(this.currentProtoView);
+    // Attention: variables present on an embedded template count towards
+    // the embedded template and not the template anchor!
+    templateVisitAll(this, ast.vars);
+    templateVisitAll(this, ast.children);
+    this.currentProtoView = prevStack;
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: pop the stack here!

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    componentType: TypeMetadata, componentStrategy: ChangeDetectionStrategy,
+    genConfig: ChangeDetectorGenConfig, parsedTemplate: TemplateAst[]): ChangeDetectorDefinition[] {
+  var visitor = new ProtoViewVisitor(componentStrategy);
+  templateVisitAll(visitor, parsedTemplate);
+  return createChangeDefinitions(visitor.protoViewStack, componentType, genConfig);
+}
+
+class ProtoViewVisitor implements TemplateAstVisitor {
+  viewCount: number = 0;
+  currentProtoView: ProtoViewVisitorData;
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: make this a getter, i.e. the top of the stack

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    this.protoViewStack.push(this.currentProtoView);
+  }
+
+  visitEmbeddedTemplate(ast: EmbeddedTemplateAst): any {
+    this.currentProtoView.boundElementCount++;
+    this.directiveRecord = null;
+    templateVisitAll(this, ast.directives);
+
+    this.viewCount++;
+    var prevStack = this.currentProtoView;
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: name is wrong

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    // Attention: variables present on an embedded template count towards
+    // the embedded template and not the template anchor!
+    templateVisitAll(this, ast.vars);
+    templateVisitAll(this, ast.children);
+    this.currentProtoView = prevStack;
+    return null;
+  }
+
+  visitElement(ast: ElementAst): any {
+    if (ast.properties.length > 0 || ast.events.length > 0 || ast.vars.length > 0 ||
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: maybe add a getter `isBound` to `ElementAst`

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        ast.directives.length > 0) {
+      this.currentProtoView.boundElementCount++;
+    }
+    this.directiveRecord = null;
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: remove this

Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: change the visitor to take an extra `context` parameter in every method

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+  visitNgContent(ast: NgContentAst): any { return null; }
+
+  visitVariable(ast: VariableAst): any {
+    this.currentProtoView.variableNames.push(ast.name);
+    return null;
+  }
+
+  visitEvent(ast: BoundEventAst): any {
+    var bindingRecord;
+    bindingRecord =
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: inline `var`

File: modules/angular2/src/compiler/change_definition_factory.ts [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+    if (this.directiveRecord.callOnChanges) {
+      bindingRecords.push(BindingRecord.createDirectiveOnChanges(this.directiveRecord));
+    }
+    if (this.directiveRecord.callOnInit) {
+      bindingRecords.push(BindingRecord.createDirectiveOnInit(this.directiveRecord));
+    }
+    if (this.directiveRecord.callDoCheck) {
+      bindingRecords.push(BindingRecord.createDirectiveDoCheck(this.directiveRecord));
+    }
+    templateVisitAll(this, ast.properties);
```
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: Order: properties first, callbacks, then host...

File: modules/angular2/src/compiler/change_definition_factory.ts (hunk 3 above)
Comment (tbosch) [inline modules/angular2/src/compiler/change_definition_factory.ts]: keep all created ProtoViewVisitorData

**Comment by tbosch** [comment]
Change `Ast` to use a context. For directives, do a manual loop and pass in the directive index.

**Comment by tbosch** [comment]
Reviewed with @vsavkin in person.

**Pull request closed without merging**